<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GPCRAnalysis.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://HolyLab.github.io/GPCRAnalysis.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GPCRAnalysis.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/master/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="GPCRAnalysis"><a class="docs-heading-anchor" href="#GPCRAnalysis">GPCRAnalysis</a><a id="GPCRAnalysis-1"></a><a class="docs-heading-anchor-permalink" href="#GPCRAnalysis" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/HolyLab/GPCRAnalysis.jl">GPCRAnalysis</a>.</p><ul><li><a href="#GPCRAnalysis.AccessionCode-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment, AbstractString}"><code>GPCRAnalysis.AccessionCode</code></a></li><li><a href="#GPCRAnalysis.BWScheme-Tuple{AbstractVector, AbstractVector}"><code>GPCRAnalysis.BWScheme</code></a></li><li><a href="#GPCRAnalysis.NWGapCosts-Union{Tuple{}, Tuple{T}} where T"><code>GPCRAnalysis.NWGapCosts</code></a></li><li><a href="#GPCRAnalysis.SequenceMapping-Tuple{MIToS.MSA.AnnotatedAlignedSequence}"><code>GPCRAnalysis.SequenceMapping</code></a></li><li><a href="#GPCRAnalysis.StructAlign-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractString}"><code>GPCRAnalysis.StructAlign</code></a></li><li><a href="#GPCRAnalysis.align-Tuple{AbstractMatrix{Float64}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, SequenceMapping}"><code>GPCRAnalysis.align</code></a></li><li><a href="#GPCRAnalysis.align_closest-Tuple{Any, Any}"><code>GPCRAnalysis.align_closest</code></a></li><li><a href="#GPCRAnalysis.align_nw-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, NWGapCosts}"><code>GPCRAnalysis.align_nw</code></a></li><li><a href="#GPCRAnalysis.align_nw-Tuple{AbstractMatrix, NWGapCosts}"><code>GPCRAnalysis.align_nw</code></a></li><li><a href="#GPCRAnalysis.align_ranges-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractVector{&lt;:BioStructures.AbstractResidue}, AbstractVector{&lt;:AbstractUnitRange}}"><code>GPCRAnalysis.align_ranges</code></a></li><li><a href="#GPCRAnalysis.align_to_axes-Tuple{Any}"><code>GPCRAnalysis.align_to_axes</code></a></li><li><a href="#GPCRAnalysis.align_to_membrane-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Any}"><code>GPCRAnalysis.align_to_membrane</code></a></li><li><a href="#GPCRAnalysis.alphacarbon_coordinates-Tuple{Any}"><code>GPCRAnalysis.alphacarbon_coordinates</code></a></li><li><a href="#GPCRAnalysis.alphacarbon_coordinates_matrix-Tuple{Any}"><code>GPCRAnalysis.alphacarbon_coordinates_matrix</code></a></li><li><a href="#GPCRAnalysis.alphafoldfile"><code>GPCRAnalysis.alphafoldfile</code></a></li><li><a href="#GPCRAnalysis.alphafoldfiles"><code>GPCRAnalysis.alphafoldfiles</code></a></li><li><a href="#GPCRAnalysis.alphafoldfiles"><code>GPCRAnalysis.alphafoldfiles</code></a></li><li><a href="#GPCRAnalysis.chargelocations-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}}"><code>GPCRAnalysis.chargelocations</code></a></li><li><a href="#GPCRAnalysis.chimerax_script-Tuple{Any, Any, MIToS.MSA.AnnotatedMultipleSequenceAlignment, Any}"><code>GPCRAnalysis.chimerax_script</code></a></li><li><a href="#GPCRAnalysis.columnwise_entropy"><code>GPCRAnalysis.columnwise_entropy</code></a></li><li><a href="#GPCRAnalysis.download_alphafolds-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.download_alphafolds</code></a></li><li><a href="#GPCRAnalysis.features_from_structure"><code>GPCRAnalysis.features_from_structure</code></a></li><li><a href="#GPCRAnalysis.filter_long!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, Real}"><code>GPCRAnalysis.filter_long!</code></a></li><li><a href="#GPCRAnalysis.filter_species!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, AbstractString}"><code>GPCRAnalysis.filter_species!</code></a></li><li><a href="#GPCRAnalysis.forcecomponents"><code>GPCRAnalysis.forcecomponents</code></a></li><li><a href="#GPCRAnalysis.forcedict-Union{Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number}, Tuple{K}, Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number, AbstractVector}} where K"><code>GPCRAnalysis.forcedict</code></a></li><li><a href="#GPCRAnalysis.getchain-Tuple{AbstractString}"><code>GPCRAnalysis.getchain</code></a></li><li><a href="#GPCRAnalysis.inward_ecl_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}"><code>GPCRAnalysis.inward_ecl_residues</code></a></li><li><a href="#GPCRAnalysis.inward_tm_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}"><code>GPCRAnalysis.inward_tm_residues</code></a></li><li><a href="#GPCRAnalysis.lookupbw-Tuple{Integer, BWScheme}"><code>GPCRAnalysis.lookupbw</code></a></li><li><a href="#GPCRAnalysis.map_closest-Tuple{Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}}"><code>GPCRAnalysis.map_closest</code></a></li><li><a href="#GPCRAnalysis.map_uniprot_retrieve-Tuple{Any}"><code>GPCRAnalysis.map_uniprot_retrieve</code></a></li><li><a href="#GPCRAnalysis.map_uniprot_status-Tuple{Any}"><code>GPCRAnalysis.map_uniprot_status</code></a></li><li><a href="#GPCRAnalysis.map_uniprot_submit"><code>GPCRAnalysis.map_uniprot_submit</code></a></li><li><a href="#GPCRAnalysis.marker-Tuple{Integer, AbstractVector{&lt;:Real}, Real, Any}"><code>GPCRAnalysis.marker</code></a></li><li><a href="#GPCRAnalysis.optimize_weights-Tuple{AbstractVector{&lt;:AbstractMatrix}}"><code>GPCRAnalysis.optimize_weights</code></a></li><li><a href="#GPCRAnalysis.pLDDTcolor-Tuple{BioStructures.Residue}"><code>GPCRAnalysis.pLDDTcolor</code></a></li><li><a href="#GPCRAnalysis.project_sequences-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.project_sequences</code></a></li><li><a href="#GPCRAnalysis.query_alphafold_latest-Tuple{AbstractString}"><code>GPCRAnalysis.query_alphafold_latest</code></a></li><li><a href="#GPCRAnalysis.query_ebi_proteins-Tuple{Any}"><code>GPCRAnalysis.query_ebi_proteins</code></a></li><li><a href="#GPCRAnalysis.query_ncbi-Tuple{Any}"><code>GPCRAnalysis.query_ncbi</code></a></li><li><a href="#GPCRAnalysis.query_uniprot_accession-Tuple{Any}"><code>GPCRAnalysis.query_uniprot_accession</code></a></li><li><a href="#GPCRAnalysis.residue_centroid-Tuple{BioStructures.AbstractResidue}"><code>GPCRAnalysis.residue_centroid</code></a></li><li><a href="#GPCRAnalysis.residue_centroid_matrix-Tuple{Any}"><code>GPCRAnalysis.residue_centroid_matrix</code></a></li><li><a href="#GPCRAnalysis.residueindex-Tuple{StructAlign, Integer, Nothing}"><code>GPCRAnalysis.residueindex</code></a></li><li><a href="#GPCRAnalysis.residueindex-Tuple{StructAlign, Nothing, Integer}"><code>GPCRAnalysis.residueindex</code></a></li><li><a href="#GPCRAnalysis.sortperm_msa-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.sortperm_msa</code></a></li><li><a href="#GPCRAnalysis.species-Tuple{AbstractString}"><code>GPCRAnalysis.species</code></a></li><li><a href="#GPCRAnalysis.try_download_alphafold"><code>GPCRAnalysis.try_download_alphafold</code></a></li><li><a href="#GPCRAnalysis.uniprotX-Tuple{AbstractString}"><code>GPCRAnalysis.uniprotX</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.AccessionCode-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment, AbstractString}" href="#GPCRAnalysis.AccessionCode-Tuple{MIToS.MSA.AnnotatedMultipleSequenceAlignment, AbstractString}"><code>GPCRAnalysis.AccessionCode</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ac = AccessionCode(msa, seqname)</code></pre><p>Return the Uniprot accession code associated with <code>seqname</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/msa.jl#L45-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.BWScheme-Tuple{AbstractVector, AbstractVector}" href="#GPCRAnalysis.BWScheme-Tuple{AbstractVector, AbstractVector}"><code>GPCRAnalysis.BWScheme</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">BWScheme(conserved_idx, tmspans)</code></pre><p>Specify the Ballesteros-Weinstein scheme used for a particular protein. <code>conserved_idx</code> is a list of 7 &quot;most conserved&quot; residues per helix (rhodopsin family: N1, D2, R3, W4, P5, P6, P7) and the span of each helix.</p><p><strong>Examples</strong></p><p>For mouse rhodopsin (P15409),</p><pre><code class="language-julia hljs">julia&gt; opsd_scheme = BWScheme([55, 83, 135, 161, 215, 267, 303],
                              [37:61, 74:96, 111:133, 153:173, 203:224, 253:274, 287:308]);</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/ballesteros_weinstein.jl#L6-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.NWGapCosts-Union{Tuple{}, Tuple{T}} where T" href="#GPCRAnalysis.NWGapCosts-Union{Tuple{}, Tuple{T}} where T"><code>GPCRAnalysis.NWGapCosts</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">gapcosts = NWGapCosts{T}(; extend1=0, extend2=0, open1=0, open2=0)</code></pre><p>Create an affine cost for gaps in Needleman-Wunsch alignment. The cost of a gap of length <code>k</code> is</p><pre><code class="nohighlight hljs">extend * k + open</code></pre><p>All costs must be nonnegative.</p><p><code>gapcosts(ϕ, idxs1, idxs2)</code> computes the contribution of gaps to the cost of alignment <code>ϕ</code> between two sequences with <code>idxs1 = eachindex(seq1)</code> and <code>idxs2 = eachindex(seq2)</code>. (The indices are needed to determine whether the alignment starts or ends with a gap.)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L198-L212">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.SequenceMapping-Tuple{MIToS.MSA.AnnotatedAlignedSequence}" href="#GPCRAnalysis.SequenceMapping-Tuple{MIToS.MSA.AnnotatedAlignedSequence}"><code>GPCRAnalysis.SequenceMapping</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sm = SequenceMapping([4, 5, 0, ...])
sm = SequenceMapping(seq::AnnotatedAlignedSequence)</code></pre><p>A <code>SequenceMapping</code> is a vector of indexes within a full sequence that map to a reference. Specifically, <code>sm[i]</code> is the index of the residue in the full sequence that maps to the <code>i</code>-th position in the reference. 0 is a placeholder for a position in the reference that has no mapping to the full sequence.</p><p><strong>Example</strong></p><p><code>SequenceMapping([4, 5, 0, ...])</code> indicates that:</p><ul><li>the first position in the reference maps to the fourth residue in the full sequence,</li><li>the second position in the reference maps to the fifth residue in the full sequence, and</li><li>the third position in the reference lacks a corresponding residue in the full sequence.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/msa.jl#L80-L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.StructAlign-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractString}" href="#GPCRAnalysis.StructAlign-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractString}"><code>GPCRAnalysis.StructAlign</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">StructAlign(struct1::ChainLike, struct2::ChainLike, filename::AbstractString)</code></pre><p>Create a structure-based alignment between <code>struct1</code> and <code>struct2</code>. <code>filename</code> is the name of the TM-align &quot;results&quot; file (e.g., https://zhanggroup.org//TM-align/example/873772.html).</p><p>See also <a href="#GPCRAnalysis.residueindex-Tuple{StructAlign, Integer, Nothing}"><code>residueindex</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/tmalign.jl#L60-L67">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align-Tuple{AbstractMatrix{Float64}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, SequenceMapping}" href="#GPCRAnalysis.align-Tuple{AbstractMatrix{Float64}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, SequenceMapping}"><code>GPCRAnalysis.align</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">tform = align(fixedpos::AbstractMatrix{Float64}, moving::Chain, sm::SequenceMapping)
tform = align(fixed::Chain, moving::Chain, sm::SequenceMapping)</code></pre><p>Return a rotated and shifted version of <code>moving</code> so that the centroids of residues <code>moving[sm]</code> have least mean square error deviation from positions <code>fixedpos</code> or those of residues <code>fixed</code>. <code>fixed</code> or <code>fixedpos</code> should include just the residues or positions you want to align to, but <code>moving</code> should be an entire chain.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L3-L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_closest-Tuple{Any, Any}" href="#GPCRAnalysis.align_closest-Tuple{Any, Any}"><code>GPCRAnalysis.align_closest</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">tform = align_closest(mapto::StructLike, mapfrom::StructLike; Dthresh=5)
tform = align_closest(coordsto, coordsfrom; Dthresh=5)</code></pre><p>Return the rigid transformation best aligning <code>mapfrom</code> to <code>mapto</code>. The transformation is computed from residues matched by <code>map_closest</code>, using only residues closer than <code>Dthresh</code>.</p><p>Because the mapping is determined by distance, this can only &quot;tweak&quot; an already-close alignment.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L50-L58">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_nw-Tuple{AbstractMatrix, NWGapCosts}" href="#GPCRAnalysis.align_nw-Tuple{AbstractMatrix, NWGapCosts}"><code>GPCRAnalysis.align_nw</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ϕ = align_nw(D, gapcosts::NWGapCosts)</code></pre><p>Given a pairwise penalty matrix <code>D</code> (e.g., a pairwise distance matrix) and costs for opening and extending gaps, find the optimal pairings</p><pre><code class="nohighlight hljs">ϕ = [(i1, j1), (i2, j2), ...]</code></pre><p>that minimize</p><pre><code class="nohighlight hljs">sum(D[ϕk...] for ϕk in ϕ) + gapcosts(ϕ, axes(D)...)</code></pre><p>subject to the constraint that <code>all(ϕ[k+1] .&gt; ϕ[k])</code> for all <code>k</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L232-L245">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_nw-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, NWGapCosts}" href="#GPCRAnalysis.align_nw-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, NWGapCosts}"><code>GPCRAnalysis.align_nw</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ϕ = align_nw(seq1, seq2, gapcosts::NWGapCosts; mode=:distance_orientation)</code></pre><p>Find the optimal <code>ϕ</code> matching <code>seq1[ϕ[k][1]]</code> to <code>seq2[ϕ[k][2]]</code> for all <code>k</code>. <code>mode</code> controls the computation of pairwise matching penalties, and can be either <code>:distance</code> or <code>:distance_orientation</code>, where the latter adds any mismatch in sidechain orientation to the distance penalty.</p><p><code>seq1</code> and <code>seq2</code> must be aligned to each other in 3D space before calling this function. See <a href="#GPCRAnalysis.align-Tuple{AbstractMatrix{Float64}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, SequenceMapping}"><code>align</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L252-L262">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_ranges-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractVector{&lt;:BioStructures.AbstractResidue}, AbstractVector{&lt;:AbstractUnitRange}}" href="#GPCRAnalysis.align_ranges-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, AbstractVector{&lt;:BioStructures.AbstractResidue}, AbstractVector{&lt;:AbstractUnitRange}}"><code>GPCRAnalysis.align_ranges</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">seqtms = align_ranges(seq1, seq2, seq2ranges::AbstractVector{&lt;:AbstractUnitRange})</code></pre><p>Transfer <code>refranges</code>, a list of reside index spans in <code>seq2</code>, to <code>seq1</code>. <code>seq1</code> and <code>seq2</code> must be spatially aligned, and the assignment is made by minimizing inter-chain distance subject to the constraint of preserving sequence order.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L279-L285">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_to_axes-Tuple{Any}" href="#GPCRAnalysis.align_to_axes-Tuple{Any}"><code>GPCRAnalysis.align_to_axes</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">tform = align_to_axes(strct)</code></pre><p>Compute the transformation needed to align the principle axes of inertia of <code>strct</code> with the coordinate axes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L141-L146">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.align_to_membrane-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Any}" href="#GPCRAnalysis.align_to_membrane-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Any}"><code>GPCRAnalysis.align_to_membrane</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">tform = align_to_membrane(chain::ChainLike, tms; extracellular=true)</code></pre><p>Compute the rigid transformation <code>tform</code> needed to align <code>chain</code> to the membrane, given the transmembrane segments <code>tms</code> as residue-indexes (e.g., <code>[37:61, 74:96, ...]</code>). <code>extracelluar</code> should be true if the N-terminus of the protein is extracellular (<code>chain[first(tms[1])]</code> is at the extracellular face), and false otherwise.</p><p><code>applytransform!(chain, tform)</code> (or the <code>model</code> that includes <code>chain</code>) will re-orient <code>chain</code> so that the center of the membrane is <code>z=0</code> and extracellular is positive.</p><p>The algorithm finds the membrane normal <code>u</code> by maximizing the ratio</p><pre><code class="nohighlight hljs">sumᵢ (u ⋅ vᵢ)²</code></pre><hr/><pre><code class="nohighlight hljs">sumᵢ (u ⋅ δᵢ)²</code></pre><p>where <code>vᵢ</code> is a vector parallel to the <code>i</code>th TM helix, and <code>δᵢ</code> is a within-leaflet atomic displacement.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L80-L101">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.alphacarbon_coordinates-Tuple{Any}" href="#GPCRAnalysis.alphacarbon_coordinates-Tuple{Any}"><code>GPCRAnalysis.alphacarbon_coordinates</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">alphacarbon_coordinates(res::AbstractResidue)</code></pre><p>Return the coordinates of the α-carbon in <code>res</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L63-L67">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.alphacarbon_coordinates_matrix-Tuple{Any}" href="#GPCRAnalysis.alphacarbon_coordinates_matrix-Tuple{Any}"><code>GPCRAnalysis.alphacarbon_coordinates_matrix</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">alphacarbon_coordinates_matrix(seq)</code></pre><p>Return a matrix of αC coordinates as columns across all residues. See also <a href="#GPCRAnalysis.alphacarbon_coordinates-Tuple{Any}"><code>alphacarbon_coordinates</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L70-L74">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.alphafoldfile" href="#GPCRAnalysis.alphafoldfile"><code>GPCRAnalysis.alphafoldfile</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fns = alphafoldfile(uniprotXname, dirname=pwd(); join=false)</code></pre><p>Return the latest version of the AlphaFold file for <code>uniprotXname</code> in <code>dirname</code>. If <code>join</code> is <code>true</code>, then the full path is returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L8-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.alphafoldfiles" href="#GPCRAnalysis.alphafoldfiles"><code>GPCRAnalysis.alphafoldfiles</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fns = alphafoldfiles(dirname=pwd(); join=false)</code></pre><p>Return the latest version of all AlphaFold files in <code>dirname</code>. If <code>join</code> is <code>true</code>, then the full paths are returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L29-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.alphafoldfiles" href="#GPCRAnalysis.alphafoldfiles"><code>GPCRAnalysis.alphafoldfiles</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">msacode2structfile = alphafoldfiles(msa::AnnotatedMultipleSequenceAlignment, dirname=pwd())</code></pre><p>Return a dictionary mapping <code>MSACode</code>s to the corresponding AlphaFold structure files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L53-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.chargelocations-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}}" href="#GPCRAnalysis.chargelocations-Tuple{Union{BioStructures.Chain, AbstractVector{&lt;:BioStructures.AbstractResidue}}}"><code>GPCRAnalysis.chargelocations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">chargelocations(chain::ChainLike; include_his::Bool=false)</code></pre><p>Return a list of potential charge locations in the protein structure. Each is a tuple <code>(position, residueindex, AAname)</code>. The positions are those of <code>N</code> (in positively-charged residues like Arg &amp; Lys) or <code>O</code> (in negatively-charged residues like Asp and Glu). N- and C-termini are not included in the list. While each residue will carry a net total charge of ±1, the location of each potential charge will be listed (1 for Lys, 2 each for Arg, Asp, and Glu).</p><p>By default, histidine is not considered charged, but you can include it by setting <code>include_His=true</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L77-L86">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.chimerax_script-Tuple{Any, Any, MIToS.MSA.AnnotatedMultipleSequenceAlignment, Any}" href="#GPCRAnalysis.chimerax_script-Tuple{Any, Any, MIToS.MSA.AnnotatedMultipleSequenceAlignment, Any}"><code>GPCRAnalysis.chimerax_script</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">chimerax_script(scriptfilename, uprot_list, msa::AnnotatedMultipleSequenceAlignment, colidxs;
                dir=pwd(), align=true, chain_transparency=80, styles=Dict{Int,String}(), extras=String[])</code></pre><p>Create a <a href="https://www.cgl.ucsf.edu/chimerax/">chimerax</a> visualization script with name <code>scriptfilename</code>. <code>uprot_list</code> is a list of UniProtX names that you want to visualize. <code>msa</code> is a Multiple Sequence alignment and <code>colidxs</code> specifies the column indices in <code>msa</code> corresponding to amino acid side chains that you&#39;d like to visualize.</p><p>Keyword arguments:</p><ul><li><code>dir</code> is the directory with the protein structure files</li><li><code>align</code> determines whether to align the structures to the first one (uses the <code>matchmaker</code> tool)</li><li><code>chain_transparency</code> sets the transparency on the ribbon diagrams (0 = not</li></ul><p>transparent)</p><ul><li><code>styles</code> can be used to affect the display, e.g., <code>Dict(k =&gt; &quot;@SD sphere&quot;)</code></li></ul><p>would cause methionines at column index <code>k</code> to be displayed with the sulfur in sphere mode.</p><ul><li><code>extras</code> can be used to hand-specify a number of additional commands; this can</li></ul><p>be useful if, for example, the <code>msa</code> has occasional misalignments.</p><p><strong>Examples</strong></p><p>Suppose you have the <code>msa</code> for rhodopsin (mouse: P15409), then:</p><pre><code class="language-julia hljs">chimerax_script(&quot;myscript.cxc&quot;, [&quot;P15409&quot;], msa, [i1, i2, i3])</code></pre><p>where <code>i1</code> through <code>i3</code> are column-indices in the <code>msa</code> that you&#39;d like to view.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/chimerax.jl#L34-L65">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.columnwise_entropy" href="#GPCRAnalysis.columnwise_entropy"><code>GPCRAnalysis.columnwise_entropy</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">columnwise_entropy(msa::AbstractMultipleSequenceAlignment, aacode = ReducedAlphabet(&quot;(AILMV)(NQST)(RHK)(DE)(FWY)CGP&quot;))</code></pre><p>Compute the entropy of each column in an MSA. Low entropy indicates high conservation.</p><p>Unmatched entries (<code>&#39;-&#39;</code> residues) contribute to the entropy calculation as if they were an ordinary residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L32-L38">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.download_alphafolds-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}" href="#GPCRAnalysis.download_alphafolds-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.download_alphafolds</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">download_alphafolds(msa::AbstractMultipleSequenceAlignment; dirname=pwd())
download_alphafolds(ids; dirname=pwd())</code></pre><p>Download all available <a href="https://alphafold.com/">AlphaFold</a> structures for the sequences in <code>msa</code>. Missing entries are silently skipped.</p><p>If a <code>AbstractMultipleSequenceAlignment</code> is provided, the downloaded PDB file is checked to ensure that the residues in the MSA sequence match those in the PDB file. If they do not match, the PDB file is removed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L118-L127">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.features_from_structure" href="#GPCRAnalysis.features_from_structure"><code>GPCRAnalysis.features_from_structure</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mgmm = features_from_structure(seq::ChainLike, idxs=1:length(seq); combined=false)</code></pre><p>Construct an <code>IsotropicMultiGMM</code> from <code>seq</code> by adding features for each residue in <code>idxs</code>.</p><p><code>combined=true</code> causes all atoms sharing the same feature to be combined, reducing the total number of features in the resulting model. The default creates features for each atom separately.</p><p>The <code>σfun</code> and <code>ϕfun</code> keyword arguments are functions that determine the standard deviation and amplitude of each gaussian feature, respectively, and take arguments <code>(atom, residue, feature)</code>.</p><p>The output <code>mgmm</code> may include the following features:</p><ul><li><code>:Steric</code> (the &quot;hard center&quot;, a proxy for the repulsive core of Lennard-Jones potentials)</li><li><code>:Hydrophobe</code> (van der Waals interactions, a proxy for the attractive part of Lennard-Jones potentials)</li><li><code>:Aromatic</code> (the aromatic ring of phenylalanine, tyrosine, and tryptophan)</li><li><code>:PosIonizable</code> (the positively charged nitrogen of histidine, arginine, and lysine; histidine gets a fractional charge of +0.1)</li><li><code>:NegIonizable</code> (the negatively charged oxygen of aspartate and glutamate; each oxygen gets a fractional charge of -0.5)</li><li><code>:Donor</code> (the hydrogen of a hydrogen bond donor)</li><li><code>:Acceptor</code> (the oxygen of a hydrogen bond acceptor)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/features.jl#L124-L146">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.filter_long!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, Real}" href="#GPCRAnalysis.filter_long!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, Real}"><code>GPCRAnalysis.filter_long!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">filter_long!(msa::AbstractMultipleSequenceAlignment, minres::Real)</code></pre><p>Remove all sequences from <code>msa</code> with fewer than <code>minres</code> matching residues.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/msa.jl#L31-L35">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.filter_species!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, AbstractString}" href="#GPCRAnalysis.filter_species!-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment, AbstractString}"><code>GPCRAnalysis.filter_species!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">filter_species!(msa::AbstractMultipleSequenceAlignment, speciesname::AbstractString)</code></pre><p>Remove all sequences from <code>msa</code> except those with <a href="#GPCRAnalysis.species-Tuple{AbstractString}"><code>species(sequencename)</code></a> equal to <code>speciesname</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/msa.jl#L21-L25">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.forcecomponents" href="#GPCRAnalysis.forcecomponents"><code>GPCRAnalysis.forcecomponents</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">forces = forcecomponents(seq, interactions::AbstractVector, residueindexes=eachindex(seq); kwargs...)</code></pre><p>Calculate the forces between residues in <code>seq</code> based on the features of each residue and the given <code>interactions</code>, which must be a list of 2-tuples <code>(:field1, :field2)</code> or pairs <code>(:field1, :field2) =&gt; coef</code>, where <code>:field1</code> and <code>:field2</code> are the names of the features that interact and <code>coef</code> is the coefficient of the force (defaults to 1). The optional keyword arguments <code>kwargs</code> are as described in <a href="#GPCRAnalysis.features_from_structure"><code>features_from_structure</code></a>.</p><p>The feature-names can be the ones used in <a href="#GPCRAnalysis.features_from_structure"><code>features_from_structure</code></a>. Each pair of interactions should be listed, but only in one order (symmetry is automatically enforced). Optionally, you can also &quot;bundle&quot; features together: <code>:Ionic</code> is a bundle of <code>:PosIonizable</code> and <code>:NegIonizable</code>, where like charges repel and opposite charges attract with the same magnitude of force. Using <code>(:Ionic,:Ionic)</code> instead of listing all three interactions (<code>:PosIonizable,:PosIonizable</code>, <code>:NegIonizable,:NegIonizable</code>, and <code>:PosIonizable,:NegIonizable</code>) separately ensures that any tuning of ionic forces will satisfy the symmetries of the real world.</p><p>Upon return, there is one force-matrix for each residue in <code>seq</code> listed in <code>residueindexes</code>. Each force-matrix is a <code>3×n</code> matrix where each row corresponds to a force component <code>(x, y, z)</code> and the <code>k</code>th column corresponds to <code>interactions[k]</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia hljs">julia&gt; seq = getchain(&quot;1GZM.pdb&quot;)
interactions = [(:Steric, :Steric) =&gt; 1,           # repulsive
                (:Hydrophobe, :Hydrophobe) =&gt; -1,  # attractive
                (:Donor, :Acceptor) =&gt; -1          # attractive
                ]
julia&gt; forces = forcecomponents(seq, interactions)</code></pre><p>The output is a list of 3×4 matrices, one for each residue in <code>seq</code>.</p><p>See also: <a href="#GPCRAnalysis.optimize_weights-Tuple{AbstractVector{&lt;:AbstractMatrix}}"><code>optimize_weights</code></a>, <a href="#GPCRAnalysis.forcedict-Union{Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number}, Tuple{K}, Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number, AbstractVector}} where K"><code>forcedict</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/forces.jl#L5-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.forcedict-Union{Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number}, Tuple{K}, Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number, AbstractVector}} where K" href="#GPCRAnalysis.forcedict-Union{Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number}, Tuple{K}, Tuple{AbstractArray{Pair{Tuple{K, K}, T}, 1} where T&lt;:Number, AbstractVector}} where K"><code>GPCRAnalysis.forcedict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">interactiondict = forcedict(interactions::AbstractVector, w = ones(length(interactions)))</code></pre><p>Create a dictionary of interactions, where <code>interactions[i]</code> should be weighted by <code>w[i]</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/forces.jl#L145-L149">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.getchain-Tuple{AbstractString}" href="#GPCRAnalysis.getchain-Tuple{AbstractString}"><code>GPCRAnalysis.getchain</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">getchain(filename::AbstractString; model=1, chain=&quot;A&quot;)</code></pre><p>Read a PDB or mmCIF file <code>filename</code> and extract the specified chain.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/utils.jl#L40-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.inward_ecl_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}" href="#GPCRAnalysis.inward_ecl_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}"><code>GPCRAnalysis.inward_ecl_residues</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">inward_ecl_residues(seq, eclidxs)</code></pre><p>Return an array of boolean[] indicating which residues (of those specified by <code>eclidxs</code>) are inward-facing (i.e. downward toward the opening of the binding pocket).</p><p><code>eclidxs</code> is a vector (with each entry corresponding to an extracellular loop) of ranges of residue indices.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/pocket.jl#L65-L71">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.inward_tm_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}" href="#GPCRAnalysis.inward_tm_residues-Tuple{AbstractVector{&lt;:BioStructures.AbstractResidue}, Any}"><code>GPCRAnalysis.inward_tm_residues</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">inward_tm_residues(seq, tmidxs)</code></pre><p>Return an array of boolean[] indicating which residues (of those specified by <code>tmidxs</code>) are inward-facing.</p><p><code>tmidxs</code> is a vector (typically of length 7, with each entry corresponding to a transmembrane region) of ranges of residue indices.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/pocket.jl#L43-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.lookupbw-Tuple{Integer, BWScheme}" href="#GPCRAnalysis.lookupbw-Tuple{Integer, BWScheme}"><code>GPCRAnalysis.lookupbw</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">helix, residue_position = lookupbw(idx::Integer, scheme::BWScheme)</code></pre><p>Calculate the Ballesteros-Weinstein residue number coresponding to residue <code>idx</code>. <code>tmspans</code> describes the transmembrane regions in the reference, and <code>bwconserved</code> the index of the most-conserved residue.</p><p><strong>Examples</strong></p><p>For mouse rhodopsin (P15409),</p><pre><code class="language-julia hljs">julia&gt; opsd_scheme = BWScheme([55, 83, 135, 161, 215, 267, 303],
           [34:64, 73:99, 107:139, 150:173, 200:229, 246:277, 285:309]);

julia&gt; lookupbw(160, opsd_scheme)
(4, 49)

julia&gt; lookupbw((4, 49), opsd_scheme)
160</code></pre><p>This is the residue just before the most-conserved residue of helix 4.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/ballesteros_weinstein.jl#L27-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.map_closest-Tuple{Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}}" href="#GPCRAnalysis.map_closest-Tuple{Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}, Union{BioStructures.Chain, BioStructures.Model, BioStructures.MolecularStructure, AbstractVector{&lt;:BioStructures.AbstractResidue}}}"><code>GPCRAnalysis.map_closest</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">mapping = map_closest(mapto::StructureLike, mapfrom::StructureLike)</code></pre><p>Return a vector <code>mapping[i] = (j, distij)</code>, matching the <code>i</code>th residue in <code>mapto</code> to the <code>j</code>th residue in <code>mapfrom</code> and reporting the distance between them. The mapping minimizes the sum of distances. <code>mapto</code> and <code>mapfrom</code> must already be aligned for this to be meaningful.</p><p>If <code>mapfrom</code> is shorter than <code>mapto</code>, some <code>j</code>s will be 0, indicating a skipped residue in <code>mapto</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/align.jl#L26-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.map_uniprot_retrieve-Tuple{Any}" href="#GPCRAnalysis.map_uniprot_retrieve-Tuple{Any}"><code>GPCRAnalysis.map_uniprot_retrieve</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">result = GPCRAnalysis.map_uniprot_retrieve(jobID)</code></pre><p>Retrieve the results of a Uniprot ID mapping job.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L145-L149">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.map_uniprot_status-Tuple{Any}" href="#GPCRAnalysis.map_uniprot_status-Tuple{Any}"><code>GPCRAnalysis.map_uniprot_status</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">status = GPCRAnalysis.map_uniprot_status(jobID)</code></pre><p>Check the status of a Uniprot ID mapping job. Returns <code>true</code> if the results are ready. Otherwise, returns the status object.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L129-L134">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.map_uniprot_submit" href="#GPCRAnalysis.map_uniprot_submit"><code>GPCRAnalysis.map_uniprot_submit</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">jobID = GPCRAnalysis.map_uniprot_submit(ids, from=&quot;UniProtKB_AC-ID&quot;, to=&quot;UniProtKB&quot;)</code></pre><p>Submit a list of <code>ids</code> to the Uniprot ID mapping service, to convert from ID convention <code>from</code> to <code>to</code>. The jobID can be used to check the status (<code>map_uniprot_status</code>) and retrieve the results (<code>map_uniprot_retrieve</code>).</p><p><strong>Examples</strong></p><p><code>jldoctest` julia&gt; jobID = GPCRAnalysis.map_uniprot_submit([&quot;ENSMUSG00000067064&quot;, &quot;ENSMUSG00000057464&quot;], &quot;Ensembl&quot;);</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L108-L118">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.marker-Tuple{Integer, AbstractVector{&lt;:Real}, Real, Any}" href="#GPCRAnalysis.marker-Tuple{Integer, AbstractVector{&lt;:Real}, Real, Any}"><code>GPCRAnalysis.marker</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">str = marker(modelnum, pos, radius, color)</code></pre><p>Return a string that represents a marker in ChimeraX. <code>modelnum</code> is the model number, <code>pos</code> is a 3D position, <code>radius</code> is the radius of the marker, and <code>color</code> is the color of the marker.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/chimerax.jl#L93-L98">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.optimize_weights-Tuple{AbstractVector{&lt;:AbstractMatrix}}" href="#GPCRAnalysis.optimize_weights-Tuple{AbstractVector{&lt;:AbstractMatrix}}"><code>GPCRAnalysis.optimize_weights</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">w = optimize_weights(forces)</code></pre><p>Tune the weights <code>w</code> to approximately satisfy force balance, i.e., solve</p><pre><code class="nohighlight hljs">min w.r.t. w  sum(sum(abs2, f*w) for f in forces)
subject to    sum(w) == 1, w .&gt;= 0</code></pre><p>This is based on the notion that the protein structure is presumably at an energy minimum.</p><p>This seems to work best for ubiquitous interactions, like <code>:Steric</code>, <code>:Hydrophobe</code>, and hydrogen-bonding. Rarer interactions (<code>:Ionic</code>, <code>:Aromatic</code>) may need to be tuned via different principles.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>This function requires that you manually load JuMP and HiGHS, e.g., <code>using JuMP, HiGHS</code>.</p></div></div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/forces.jl#L124-L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.pLDDTcolor-Tuple{BioStructures.Residue}" href="#GPCRAnalysis.pLDDTcolor-Tuple{BioStructures.Residue}"><code>GPCRAnalysis.pLDDTcolor</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pLDDTcolor(r::Residue)
pLDDTcolor(score::Real)</code></pre><p>Return the color corresponding to the pLDDT score (a measure of confidence) of a residue.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L178-L183">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.project_sequences-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}" href="#GPCRAnalysis.project_sequences-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.project_sequences</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">X = project_sequences(msa::AbstractMultipleSequenceAlignment; fracvar::Real = 0.9)</code></pre><p>Perform a classical multidimensional scaling analysis to project the sequences in <code>msa</code> to a space in which pairwise distances approximately reproduce <code>100 - percentsimilarity(seq1, seq2)</code>. The dimensionality is chosen to reconstruction <code>fracvar</code> of the variance.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.query_alphafold_latest-Tuple{AbstractString}" href="#GPCRAnalysis.query_alphafold_latest-Tuple{AbstractString}"><code>GPCRAnalysis.query_alphafold_latest</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">url = query_alphafold_latest(uniprotXname; format=&quot;cif&quot;)</code></pre><p>Query the <a href="https://alphafold.com/">AlphaFold</a> API for the latest structure of <code>uniprotXname</code>. <code>format</code> should be &quot;cif&quot;, &quot;pdb&quot;, or &quot;bcif&quot;.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L76-L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.query_ebi_proteins-Tuple{Any}" href="#GPCRAnalysis.query_ebi_proteins-Tuple{Any}"><code>GPCRAnalysis.query_ebi_proteins</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">result = query_ebi_proteins(id)</code></pre><p>Query the EBI Proteins API for a protein with the specified <code>id</code>, which must be the Uniprot accession code. You can also supply several proteins as a comma-separated list.</p><p><code>result</code> is a JSON3 object with many fields.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/utils.jl#L1-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.query_ncbi-Tuple{Any}" href="#GPCRAnalysis.query_ncbi-Tuple{Any}"><code>GPCRAnalysis.query_ncbi</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">result = query_ncbi(id)</code></pre><p>Query the NCBI API for a gene with the specified <code>id</code>, which must be the gene accession code. <code>result</code> is a JSON3 object with many fields.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/utils.jl#L25-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.query_uniprot_accession-Tuple{Any}" href="#GPCRAnalysis.query_uniprot_accession-Tuple{Any}"><code>GPCRAnalysis.query_uniprot_accession</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">accession_code = query_uniprot_accession(id)</code></pre><p>Perform a Uniprot search for <code>id</code>, returning the canonical accession code.</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; query_uniprot_accession(&quot;T2R38_MOUSE&quot;)
&quot;Q7TQA6&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L77-L88">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.residue_centroid-Tuple{BioStructures.AbstractResidue}" href="#GPCRAnalysis.residue_centroid-Tuple{BioStructures.AbstractResidue}"><code>GPCRAnalysis.residue_centroid</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">residue_centroid(r::AbstractResidue)</code></pre><p>Compute the mean position of all atoms in <code>r</code> excluding hydrogen.</p><p>Residue centers may yield a more reliable measure of &quot;comparable residues&quot; than the α-carbons (see <code>CAmatrix</code>) because they incorporate the orientation of the side chain relative to the overall fold.</p><p>See also <a href="#GPCRAnalysis.residue_centroid_matrix-Tuple{Any}"><code>residue_centroid_matrix</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L44-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.residue_centroid_matrix-Tuple{Any}" href="#GPCRAnalysis.residue_centroid_matrix-Tuple{Any}"><code>GPCRAnalysis.residue_centroid_matrix</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">residue_centroid_matrix(seq)</code></pre><p>Return a matrix of all residue centroids as columns. See also <a href="#GPCRAnalysis.residue_centroid-Tuple{BioStructures.AbstractResidue}"><code>residue_centroid</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/analyze.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.residueindex-Tuple{StructAlign, Integer, Nothing}" href="#GPCRAnalysis.residueindex-Tuple{StructAlign, Integer, Nothing}"><code>GPCRAnalysis.residueindex</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">residueindex(sa::StructAlign, idx1, nothing)
residueindex(sa::StructAlign, idx1, nothing, ±1)</code></pre><p>Calculate the residue index in structure 2 corresponding to residue <code>idx1</code> in structure 1. The second form allows you to find the nearest corresponding residue in the forward (+1) or reverse (-1) directions, if <code>idx1</code> is not among the mapped residues.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/tmalign.jl#L91-L98">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.residueindex-Tuple{StructAlign, Nothing, Integer}" href="#GPCRAnalysis.residueindex-Tuple{StructAlign, Nothing, Integer}"><code>GPCRAnalysis.residueindex</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">residueindex(sa::StructAlign, nothing, idx2)
residueindex(sa::StructAlign, nothing, idx2, ±1)</code></pre><p>Calculate the residue index in structure 1 corresponding to residue <code>idx2</code> in structure 2. The second form allows you to find the nearest corresponding residue in the forward (+1) or reverse (-1) directions, if <code>idx2</code> is not among the mapped residues.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/tmalign.jl#L71-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.sortperm_msa-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}" href="#GPCRAnalysis.sortperm_msa-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>GPCRAnalysis.sortperm_msa</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">tour = sortperm_msa(msa::AbstractMultipleSequenceAlignment)</code></pre><p>Order the sequences in <code>msa</code> to minimize the &quot;tour length&quot; visiting each sequence once. The length between sequences is defined at <code>100 - percentsimilarity(seq1, seq2)</code>.</p><p>This can be useful for graphical or alignment display by grouping obviously-similar sequences near one another.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/msa.jl#L4-L12">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.species-Tuple{AbstractString}" href="#GPCRAnalysis.species-Tuple{AbstractString}"><code>GPCRAnalysis.species</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">species(name)</code></pre><p>Extract the species identifier from a <a href="https://www.uniprot.org/help/entry_name">UniProt &quot;X_Y&quot; entry</a> or elaborated variant (e.g., PFAM sequence name).</p><p>See also <a href="#GPCRAnalysis.uniprotX-Tuple{AbstractString}"><code>uniprotX</code></a>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; species(&quot;Q8VGW6_MOUSE/31-308&quot;)
&quot;MOUSE&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L28-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.try_download_alphafold" href="#GPCRAnalysis.try_download_alphafold"><code>GPCRAnalysis.try_download_alphafold</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">try_download_alphafold(uniprotXname, path=alphafoldfilename(uniprotXname); version=4)</code></pre><p>Attempt to download an <a href="https://alphafold.com/">AlphaFold</a> structure. Returns <code>nothing</code> if no entry corresponding to <code>uniprotXname</code> exists; otherwise it returns <code>path</code>, the pathname of the saved file.</p><p>In general, a better approach is to use <a href="#GPCRAnalysis.download_alphafolds-Tuple{MIToS.MSA.AbstractMultipleSequenceAlignment}"><code>download_alphafolds</code></a> for multiple proteins, or <a href="#GPCRAnalysis.query_alphafold_latest-Tuple{AbstractString}"><code>query_alphafold_latest</code></a> combined with <code>Downloads.download</code> for a single protein.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/alphafold.jl#L92-L101">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="GPCRAnalysis.uniprotX-Tuple{AbstractString}" href="#GPCRAnalysis.uniprotX-Tuple{AbstractString}"><code>GPCRAnalysis.uniprotX</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">uniprotX(name)</code></pre><p>Extract the <a href="https://www.uniprot.org/help/entry_name">UniProt &quot;X&quot; entry</a> from an X_Y entry or elaborated variant (e.g., PFAM sequence name).</p><p>See also <a href="#GPCRAnalysis.species-Tuple{AbstractString}"><code>species</code></a>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; uniprotX(&quot;Q8VGW6_MOUSE/31-308&quot;)
&quot;Q8VGW6&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/HolyLab/GPCRAnalysis.jl/blob/3b3996fa6913a1ef1dee8ddcc386a980556ccc66/src/naming_conventions.jl#L52-L65">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.10 on <span class="colophon-date" title="Friday 15 August 2025 08:25">Friday 15 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
